sam/bamWrite.go

```diff
@@ -0,0 +1,335 @@
+
+	// BAI index bin
+	le.PutUint16(bw.u32[:2], bin)
+	bw.recordBuf.Write(bw.u32[:2])
+
+	// num cigar op
+	le.PutUint16(bw.u32[:2], uint16(len(s.Cigar)))
+	bw.recordBuf.Write(bw.u32[:2])
+
+	// flag
+	le.PutUint16(bw.u32[:2], s.Flag)
+	bw.recordBuf.Write(bw.u32[:2])
+
+	// len seq
+	le.PutUint32(bw.u32[:4], uint32(len(s.Seq)))
+	bw.recordBuf.Write(bw.u32[:4])
+
+	// next ref id
+	if s.RNext != "=" {
+		idx, ok = bw.refMap[s.RNext]
+	}
+	if !ok {
+		idx = -1
+	}
+	le.PutUint32(bw.u32[:4], uint32(idx))
+	bw.recordBuf.Write(bw.u32[:4])
+
+	// next pos
+	le.PutUint32(bw.u32[:4], s.PNext-1)
+	bw.recordBuf.Write(bw.u32[:4])
+
+	// tlen
+	le.PutUint32(bw.u32[:4], uint32(s.TLen))
+	bw.recordBuf.Write(bw.u32[:4])
+
+	// read name nul terminated
+	bw.recordBuf.WriteString(s.QName)
+	bw.recordBuf.WriteByte(nul)
+
+	// cigar
+	for i := range s.Cigar {
+		le.PutUint32(bw.u32[:4], getCigUint32(s.Cigar[i]))
+		bw.recordBuf.Write(bw.u32[:4])
+	}
+
+	// seq
+	var seqInt uint8
+	for i := range s.Seq {
+		if i%2 == 0 {
+			seqInt = baseEncoder[s.Seq[i]]
+		} else {
+			seqInt = seqInt << 4
+			seqInt = seqInt | baseEncoder[s.Seq[i]]
+			bw.recordBuf.WriteByte(seqInt)
+		}
+	}
+	if len(s.Seq)%2 != 0 { // handle odd seq len
+		seqInt = seqInt << 4
+		bw.recordBuf.WriteByte(seqInt)
+	}
+
+	// qual
+	if s.Qual == "*" {
+		for range s.Seq {
+			bw.recordBuf.WriteByte(0xff)
+		}
+	} else {
+		for i := range s.Qual {
+			bw.recordBuf.WriteByte(s.Qual[i] - 33)
+		}
+	}
+
+	// aux data
+	writeExtra(bw, s)
+
+	// write block size then write recordBuf
+	blockSize := uint32(bw.recordBuf.Len())
+	le.PutUint32(bw.u32[:4], blockSize)
+	bw.buf.Write(bw.u32[:4])           // write block size
+	bw.buf.Write(bw.recordBuf.Bytes()) // write record to buf
+
+	// Write to file once a full block has been read
+	if bw.buf.Len() >= 64000 { // Max block size is 64KB
+		n, err := bw.Write(bw.buf.Next(64000))
+		if n != 64000 || err != nil {
+			log.Panic(err)
+		}
+	}
+}
+
+// base        : =  A  C  M  G  R  S  V  T  W  Y  H  K  D  B  N
+// bam specs   : 0  1  2  3  4  5  6  7  8  9  10 11 12 13 14 15
+// gonomics val: 11 0  1  -1 2  -1 -1 -1 3  -1 -1 -1 -1 -1 -1 4
+// lowercase converted to uppercase and all non-spec bases converted to 'N'
+var baseEncoder = []uint8{1, 2, 4, 8, 15, 1, 2, 4, 8, 15, 15, 15, 15, 15, 15, 15}
+
+// getCigUint32 encodes cigar op and runlen as a uint32 defined by op_len<<4|op
+func getCigUint32(c *cigar.Cigar) uint32 {
+	var cigint uint32
+	cigint = uint32(c.RunLength) << 4  // move 4 bits to the left
+	cigint = cigint | opToUint32(c.Op) // bitwise OR with op
+	return cigint
+}
+
+// opToUint32 returns the uint32 corresponding to the rune per the Sam specifications.
+func opToUint32(r rune) uint32 {
+	switch r {
+	case 'M':
+		return 0
+	case 'I':
+		return 1
+	case 'D':
+		return 2
+	case 'N':
+		return 3
+	case 'S':
+		return 4
+	case 'H':
+		return 5
+	case 'P':
+		return 6
+	case '=':
+		return 7
+	case 'X':
+		return 8
+	default:
+		return 0 // unreachable
```

Comment:
Should this give an error if reached?

Reply:
Yes, nice catch. I must have forgotten. I put the comment there, but not the error.